File: seminar-5/doubly_linked_list/java/DoublyLinkedList.java
public class DoublyLinkedList<T> {

    private static class Node<T> {
        T value;
        Node<T> prev;
        Node<T> next;

        Node(T value, Node<T> prev, Node<T> next) {
            this.value = value;
            this.prev = prev;
            this.next = next;
        }
    }

    private Node<T> head;
    private Node<T> tail;
    private int size;

    public DoublyLinkedList() {
        head = null;
        tail = null;
        size = 0;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public void pushFront(T value) {
        Node<T> node = new Node<>(value, null, head);

        if (head == null) {
            head = tail = node;
        } else {
            head.prev = node;
            head = node;
        }
        size++;
    }

    public void pushBack(T value) {
        Node<T> node = new Node<>(value, tail, null);

        if (tail == null) {
            head = tail = node;
        } else {
            tail.next = node;
            tail = node;
        }
        size++;
    }

    public T popFront() {
        if (isEmpty()) {
            throw new IllegalStateException("pop_front from empty list");
        }

        T value = head.value;
        head = head.next;

        if (head == null) {
            tail = null;
        } else {
            head.prev = null;
        }

        size--;
        return value;
    }

    public T popBack() {
        if (isEmpty()) {
            throw new IllegalStateException("pop_back from empty list");
        }

        T value = tail.value;
        tail = tail.prev;

        if (tail == null) {
            head = null;
        } else {
            tail.next = null;
        }

        size--;
        return value;
    }

    public boolean find(T value) {
        Node<T> cur = head;
        while (cur != null) {
            if ((value == null && cur.value == null) ||
                (value != null && value.equals(cur.value))) {
                return true;
            }
            cur = cur.next;
        }
        return false;
    }

    public boolean removeFirst(T value) {
        Node<T> cur = head;

        while (cur != null) {
            boolean eq = (value == null && cur.value == null) ||
                         (value != null && value.equals(cur.value));

            if (eq) {
                if (cur.prev != null) {
                    cur.prev.next = cur.next;
                } else {
                    head = cur.next;
                }

                if (cur.next != null) {
                    cur.next.prev = cur.prev;
                } else {
                    tail = cur.prev;
                }

                size--;
                return true;
            }
            cur = cur.next;
        }
        return false;
    }

    public int size() {
        return size;
    }
}

File: seminar-5/doubly_linked_list/python/doubly_linked_list.py
class Node:
    """Узел двусвязного списка."""

    def __init__(self, value, prev=None, next=None):
        self.value = value
        self.prev = prev
        self.next = next


class DoublyLinkedList:
    """Двусвязный список."""

    def __init__(self):
        self.head = None
        self.tail = None
        self._size = 0

    def is_empty(self):
        return self._size == 0

    def push_front(self, value):
        node = Node(value, None, self.head)

        if self.head is None:
            self.head = self.tail = node
        else:
            self.head.prev = node
            self.head = node

        self._size += 1

    def push_back(self, value):
        node = Node(value, self.tail, None)

        if self.tail is None:
            self.head = self.tail = node
        else:
            self.tail.next = node
            self.tail = node

        self._size += 1

    def pop_front(self):
        if self.is_empty():
            raise IndexError("pop_front from empty list")

        value = self.head.value
        self.head = self.head.next

        if self.head is None:
            self.tail = None
        else:
            self.head.prev = None

        self._size -= 1
        return value

    def pop_back(self):
        if self.is_empty():
            raise IndexError("pop_back from empty list")

        value = self.tail.value
        self.tail = self.tail.prev

        if self.tail is None:
            self.head = None
        else:
            self.tail.next = None

        self._size -= 1
        return value

    def find(self, value):
        cur = self.head
        while cur:
            if cur.value == value:
                return True
            cur = cur.next
        return False

    def remove_first(self, value):
        cur = self.head
        while cur:
            if cur.value == value:
                if cur.prev:
                    cur.prev.next = cur.next
                else:
                    self.head = cur.next

                if cur.next:
                    cur.next.prev = cur.prev
                else:
                    self.tail = cur.prev

                self._size -= 1
                return True
            cur = cur.next
        return False

    def to_list(self):
        res = []
        cur = self.head
        while cur:
            res.append(cur.value)
            cur = cur.next
        return res

    def __len__(self):
        return self._size

    def __iter__(self):
        cur = self.head
        while cur:
            yield cur.value
            cur = cur.next

    def __repr__(self):
        return f"DoublyLinkedList(size={self._size}, values={self.to_list()})"
